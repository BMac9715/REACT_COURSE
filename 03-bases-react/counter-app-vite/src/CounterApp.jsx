import { Fragment, useState } from "react";
import PropTypes from "prop-types";

export const CounterApp = ({value}) => {

    const [ counter, setCounter ] = useState(value);

    const handleAdd = (event) => {
        //console.log(event);
        setCounter(counter + 1);
        //setCounter((c)=> c + 1);
    };

    const handleSubstract = () => {
        setCounter(counter - 1);
    }

    const resetCounter = () => {
        setCounter(value);
    }

    return (
        <Fragment>
            <h1>CounterApp</h1>
            <h2>{ counter }</h2>
            <button onClick={ handleAdd }> +1</button>
            <button onClick={ handleSubstract }> -1</button>
            <button onClick={ resetCounter }> Reset</button>
        </Fragment>
    );
}

//Props Types
CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}

//Default Props
CounterApp.defaultProps = {
    value: 0
}