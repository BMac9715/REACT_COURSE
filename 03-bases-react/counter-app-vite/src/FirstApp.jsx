import { Fragment } from "react";
import PropTypes from "prop-types";

//Esta forma, incluye un Div que esta de mas, pero en este caso es necesario para retornar el h1 y el p
export const FirstDivApp = () => {
    return (
        <div>
            <h1>First App Bryan!</h1>
            <p>Segundo tag HTML</p>
        </div>
    );
};

//Forma utilizando Fragment
export const FirstFragmentApp = () => {
    return (
        <Fragment>
            <h1>First App Bryan!</h1>
            <p>Segundo tag HTML</p>
        </Fragment>
    );
}

//Una forma sin hacer la importacion del Fragment: <> </>
export const FirstFragmentTwoApp = ( ) => {
    return (
        <>
            <h1>First App by Bryan!</h1>
            <p>Segundo tag HTML</p>
        </>
    );
}

const name = "Bryan Macario!";
const newMessage = {
    message: "Hello World",
    title: "React"
}

const getResult = (a, b) => {
    return a + b;
}

//Las props comunmente se desestructuran, no se leen como tal "props", tenerlo en cuenta
export const FirstPropsApp = ( props ) => {

    console.log(props);

    return (
        <>
            {/*<h1>First App by {name}!</h1>*/}
            <h1>{props.title}</h1>
            <h2>{getResult(5,4)}</h2>
            {/*<code>{JSON.stringify(newMessage)}</code>    Un comentario*/}
            <p>Segundo tag HTML</p>
        </>
    );
}

//Props utilizando desestructuración
export const FirstPropsDestrucApp = ( { title, subtitle } ) => {
    return (
        <>
            {/*<h1>First App by {name}!</h1>*/}
            <h1>{title}</h1>
            <p>{12 + subtitle}</p>
        </>
    );
}


export const FirstApp = ( { title, subtitle } ) => {

    /*
    if(!title) {
        throw new Error("El titulo no existe");
    }
    */

    return (
        <>
            <h1>{title}</h1>
            <p>{12 + subtitle}</p>
        </>
    );
}

//El estandar es que los Props Types y Default Props se coloquen despues de definir la función / componente

//Props Types
FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.number.isRequired
}

//Default Props
FirstApp.defaultProps = {
    title: "No hay título",
    subtitle: 0
}