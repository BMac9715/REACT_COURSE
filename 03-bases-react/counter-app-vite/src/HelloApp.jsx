export const App = () => { return (<h2>Hello World! Welcome to ReactJS!!</h2>) }

export const HelloApp = () => (<h2>Arrow Function!</h2>);