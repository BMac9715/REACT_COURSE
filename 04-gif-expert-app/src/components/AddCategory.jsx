import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ onNewCategory }) => {
    
    const [inputValue, setInputValue] = useState('Demon Slayer');

    const onInputChange = ({target}) => {
        setInputValue(target.value);
    }

    /*
    const onSubmit = (event) => {
        event.preventDefault();
        
        if(inputValue.trim().length <= 1) return;

        setCategories(cat => [...cat, inputValue]);
        setInputValue('');
    }
    */

    const onSubmit = (event) => {
        event.preventDefault();
        
        if(inputValue.trim().length <= 1) return;

        onNewCategory( inputValue.trim() );
        setInputValue('');
    }

    return (
        <form onSubmit={ onSubmit }>
            <input 
                type="text"
                placeholder="Buscar gifs..."
                value={inputValue}
                onChange={ onInputChange } />
        </form>   
  )
}

//Props Types
//AddCategory.propTypes = {
//    value: PropTypes.function.required
//}



