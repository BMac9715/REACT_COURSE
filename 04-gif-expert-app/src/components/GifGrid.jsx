import PropTypes from "prop-types";

import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
    
    const { images, isLoading } = useFetchGifs( category );

    return (
        <>
            <h3>{ category }</h3>

            {
                //isLoading ? (<h2>Cargando...</h2>) : null     //Forma 1
                isLoading && (<h2>Cargando...</h2>)             //Forma 2
            }

            <div className="card-grid">
                {
                    images.map((img) => (<GifGridItem key={img.id} { ...img }/>))
                }
            </div>      
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
