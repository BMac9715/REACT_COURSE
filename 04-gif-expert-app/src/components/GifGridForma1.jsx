import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
    
    const [images, setImages] = useState([]);

    const getImages = async () => {
        const newImages = await getGifs(category);
        setImages(newImages);
    }

    //Permite ejecutar un callback en una situación determinada
    //En este caso dejar el arreglo vacio lo interpreta como la ejecución de una sola vez
    //y esta vez es cuando se crea el componente desde cero
    useEffect(() => { 
        //Forma 1
        /*getGifs(category)
            .then( newImages => setImages(newImages) );*/

        //Forma 2
        getImages();
    }, [ ]);

    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                {
                    images.map((img) => (<GifGridItem key={img.id} { ...img }/>))
                }
            </div>      
        </>
    )
}
