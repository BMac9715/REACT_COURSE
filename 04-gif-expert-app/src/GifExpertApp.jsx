import { useState } from 'react'
import { AddCategory, GifGrid } from './components';
//import { GifGrid } from './components/GifGrid';

export const GifExpertApp = () => {
  const [categories, setCategories] = useState(['Demon Slayer']);

  const onAddCategory = (newCategory) => {
    
    if(categories.includes(newCategory)) return;

    setCategories([newCategory, ...categories]);

    //Another way
    //setCategories(cat => [...cat, 'Demon Slayer']);
  };

  return (
    <>
        {/* Titulo */}
        <h1>GifExpertApp</h1>

        {/* Input */}
        {/*<AddCategory setCategories={ setCategories }/>*/}
        <AddCategory onNewCategory={ onAddCategory }/>

        {/* Listado de Gif */}
        {
          categories.map((category) => 
            (<GifGrid key={category} category={category} />)
          )
          //categories.map(category => <GifGrid key={category} category={category} />)
        }

        <footer>
          <p>Developed by <strong>Bryan Macario</strong> guided by <strong>Fernando Herrera / Udemy Instructor</strong></p>
          <p>March 2023</p>
        </footer>
    </>
  )
}
