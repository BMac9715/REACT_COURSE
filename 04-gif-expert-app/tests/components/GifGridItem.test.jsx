import { render, screen } from "@testing-library/react";
import { GifGridItem } from "../../src/components";

describe('Prueba del componente GifGridItem', () => {
    const title = "Demon Slayer";
    const url = "https://demon-slayer.com/tanjiro.jpg"

    test('Debe hacer match con el snapshot', () => {
        const { container } = render(<GifGridItem title={title} url={url}/>);

        expect(container).toMatchSnapshot();
    });

    test('Debe mostrar la imagen con el URL y el ALT indicado', () => {
        render(<GifGridItem title={title} url={url}/>);
        //screen.debug();

        ///expect(screen.getByRole('img').src).toBe(url);
        //expect(screen.getByRole('img').alt).toBe(title);

        const {src, alt} = screen.getByRole('img');

        expect(src).toBe(url);
        expect(alt).toBe(title);
    });

    test('Debe de mostrar el titulo en el componente', () => {
        render(<GifGridItem title={title} url={url}/>);

        expect(screen.getByText(title)).toBeTruthy();
    });
});