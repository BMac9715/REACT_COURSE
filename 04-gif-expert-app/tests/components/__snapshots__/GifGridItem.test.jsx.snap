// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente GifGridItem Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Demon Slayer"
      src="https://demon-slayer.com/tanjiro.jpg"
    />
    <p>
      Demon Slayer
    </p>
  </div>
</div>
`;
